### ISSUE TITLE
monotonic attention is buggy

### ISSUE THREAD

**Issue by georgesterpu** [tensorflow/tensorflow#13669]
### System information
- **Have I written custom code **: Yes
- **OS Platform and Distribution**: Manjaro Linux, kernel 4.13.5
- **TensorFlow installed from **: binary
- **TensorFlow version (use command below)**: 1.3, 1.4 nightly (11 oct)
- **Python version**:  3.6
- **Bazel version (if compiling from source)**:
- **CUDA/cuDNN version**: cuda8, cudnn 7 & 6
- **GPU model and memory**: gtx 1080

### the problem

The two monotonic attention mechanisms, LuongMonotonicAttention and BahdanauMonotonicAttention, do not seem to work as expected on my task.

In the case of LuongMonotonicAttention, the alignment that I obtain looks like a horizontal line drawn on the first row of the image. However, the alignment has a diagonal shape using BahdanauAttention or LuongAttention, and I am instantiating these classes with the same parameters.

In the case of BahdanauMonotonicAttention, I simply receive an error message:
https://pastebin.com/raw/mPPWnEH5

Is there any preprocessing I should do in addition to the non-monotonic case ?

**Comment by craffel** [comment]
Hi, monotonic attention won't always work right out of the box when you swap softmax attention out for it.  Please see the practitioner's guide at the end of [the paper](https://arxiv.org/abs/1704.00784).  If the attention is always attending to the first encoder timestep for all decoder timesteps, you probably should try setting `score_bias_init` to a negative value.  You also will almost certainly get better results if you set `scale=True`.  If you post images of your softmax attention and the produced monotonic attention alignments, I can probably help debug further.

For the BahdanauMonotonicAttention shape error, that seems more like it may be a bug.  Can you post a full code snippet to reproduce this error?

**Comment by georgesterpu** [comment]
Here are the obtained alignments after a few iterations: [imgur link](https://imgur.com/a/R0fMe)
score_bias_init was 0 this time, just launched a new experiment with a negative value (-1)

Snippet coming soon, have to flatten the code a bit

**Comment by craffel** [comment]
> Here are the obtained alignments after a few iterations: imgur link

What are the axes?  Is decoder step vertical or horizontal?  And where is the origin?

> score_bias_init was 0 this time, just launched a new experiment with a negative value (-1)

Cool.  You may need to try progressively more negative values, and also I do recommend you try setting `scale=True`.

**Comment by georgesterpu** [comment]
Hi Colin,
The vertical axis is for the speech input, and the horizontal one for the labels. The original shapes are ~[150, 45], but I resize the images to (512,512). The origin should be the top-left corner, but I am unsure. The code is quite similar to the one in the nmt tutorial:

```
outputs, states, lengths = seq2seq.dynamic_decode(
            self._decoder_inference,
            impute_finished=True,
            swap_memory=False,
            maximum_iterations=self._hparams.max_label_length)

attention_alignment = states.alignment_history.stack()
attention_images = tf.expand_dims(tf.transpose(attention_alignment, [1, 2, 0]), -1)
attention_summary = tf.summary.image("attention_images", attention_images)
```

**Comment by craffel** [comment]
Ok, in that case monotonic attention should work - wanted to make sure you weren't reversing the input sequence, as people sometimes do in MT.  Let's get Bahdanau monotonic attention working for you since that's what I had worked with for speech.

**Comment by georgesterpu** [comment]
Turns out I had a typo in the monotonic bahdanau call, pasted the scalar `num_units` for the `memory` field too.
Indeed, the alignment images look quite different for several values of `score_bias_init` and `sigmoid_noise`, will have to search for the optimal values.
Would -1.0 and 0.5 (0.0 in test) respectively be a good starting point ? Cannot tell what would be an appropriate range right now.

**Comment by craffel** [comment]
Typically I try [-4, -3, -2, -1] for `score_bias_init`.  I have never needed to use anything other than `sigmoid_noise=1`.  Again, I recommend using `scale=True` and `BahdanauMonotonicAttention`.

**Comment by georgesterpu** [comment]
Hi Colin,
I generated some alignments using `LuongMonotonicAttention` with `mode='hard'`
https://i.imgur.com/O7PJkmp.png
The output does not look like a binary image, is this the expected behaviour ?
Same code as above to obtain the alignment images, without any resizing

**Comment by craffel** [comment]
That does not look correct for `mode='hard'`.  I'm traveling this week so will not have a chance to look at it.  Please double check that you're actually passing `mode='hard'` and then maybe post a short code snippet with how you are calling `LuongMonotonicAttention` (a minimal example showing this behavior would be most helpful).  In the meantime @ebrevdo maybe you can help debug?

**Comment by ebrevdo** [comment]
Hi George,
Are you able to provide a minimal code example that we can use to replicate
the issue?

**Comment by georgesterpu** [comment]
Thanks for the confirmation, there is probably something wrong with my code. I am also traveling tomorrow, but will check everything on Wednesday and come back with a snippet if the problem persists.
